### PR TITLE
feat(early-kickout): garbage collect ChunkProducers with block data

### PR DESCRIPTION
**Pull request by stedfn**
PR2 of a 3-PR stack for `DBCol::ChunkProducers` (nightly-only, EarlyKickout). PR1 (#16012, makes the column cold) is merged; this PR is based on master. #16046 adds cloud archival support.

## Problem

`DBCol::ChunkProducers` was never garbage collected (`GcPolicy::Other`), so it grew unbounded on long-running nodes.

## Change

Add hot GC now that PR1 made the column cold, so archival nodes copy each row to the cold DB before hot GC deletes it (no history loss).

- Flip `GcPolicy::Other` to `GcPolicy::Delete`. This and the delete calls land in one commit because `gc_col` panics on `GcPolicy::Other`.
- `clear_block_data`: delete the anchor block's rows, folded into the existing `get_shard_uids_to_gc` `ChunkExtra` loop. The key is built fresh via `get_block_shard_id(block_hash, shard_uid.shard_id())` because `ChunkProducers` is keyed by `ShardId`, not `ShardUId`.
- Header-only anchors: header sync writes a row per header, so fork siblings and headers synced ahead of their body have rows that the block-keyed `clear_block_data` never sees. Delete those in the header-GC paths by prefix-scanning `ChunkProducers` by header hash: `clear_chunk_data_and_headers` (normal canonical GC, over `chunk_tail..min_chunk_height`) and `clear_header_data_for_heights` (undo-block, over `head_height..=header_head_height`).

## Why it is safe

The consensus reader `get_chunk_producer_info_anchored` returns `ChunkProducerNotInDB` (no sampler fallback) when a same-epoch anchor row is missing. A chunk at height H anchors on its grandparent (H-2), so a row must survive while any block up to A+2 is near head.

- `clear_block_data` runs only for blocks below `gc_stop_height` (`gc_num_epochs_to_keep` epochs behind final head, computed from the body head), far below the near-head read horizon.
- The normal header-GC sweep (`clear_chunk_data_and_headers`) is bounded by `chunk_tail..min_chunk_height`, which never exceeds the tail, so it also stays far below head. Normal GC never sweeps up to `header_head`.
- The only path that deletes up to `header_head` is undo-block (`clear_header_data_for_heights` via `clear_head_block_data`), a manual rollback that discards those headers and blocks. A kept block H still finds its anchor H-2; H+2 needs H only after H is reprocessed, which rewrites the row.
- Old-fork validation after GC is unsupported and not new: `Block`, `BlockInfo`, and `ChunkExtra` for those blocks are deleted by the same pass.

## Tests (nightly)

Unit (`chain/chain/src/tests/chunk_producers.rs`):
- `test_chunk_producers_garbage_collected_with_block`: `clear_block_data` deletes a below-boundary anchor's rows while a higher anchor's rows are retained and still resolve; canonical GC clears `block_hash.prev`, not the input hash.
- `test_chunk_producers_garbage_collected_on_undo_block`: undo-block clears the body head and every header-only anchor synced above it (header_head past the body head), nothing orphaned.
- `test_chunk_producers_garbage_collected_for_header_only_fork`: normal canonical GC deletes rows for header-only hashes (fork siblings / synced-ahead headers never given a body).

Test-loop (`test-loop-tests/src/tests/chunk_producers.rs`):
- `test_chunk_producers_garbage_collected`: single-shard hot GC. A below-tail anchor's rows are deleted; an in-window anchor's rows survive and resolve; the chain advancing past the GC boundary is the deterministic stand-in for the no-stall canary.
- `test_chunk_producers_garbage_collected_at_reshard_boundary`: the boundary anchor's rows, keyed by the next (post-reshard) layout, are all deleted once it falls below the GC tail.

Green: near-chain nightly `chunk_producers` and `garbage_collection`, near-store cloud classification, test-loop `chunk_producers`, `garbage_collection`, `processed_receipts_gc`. Codex reviewed the header-path deletion for over-eviction with no findings.

## Known limitation

Cloud archival support implemented here #16046

### CODE REVIEW

File: test-loop-tests/src/tests/chunk_producers.rs
Comment (Trisfald) [inline test-loop-tests/src/tests/chunk_producers.rs]: Comment is now stale

File: chain/chain/src/garbage_collection.rs [hunk 1]
```diff
@@ -511,13 +511,30 @@ impl<'a> ChainStoreUpdate<'a> {
                 let key: &[u8] = header_hash.as_bytes();
                 store_update.delete(DBCol::BlockHeader, key);
                 self.merge(store_update);
+
+                // A ChunkProducers row is keyed by (block_hash, shard_id); the header-hash
+                // prefix matches every shard's row for this header.
+                #[cfg(feature = "nightly")]
```
Comment (Trisfald) [inline chain/chain/src/garbage_collection.rs]: nit: extract this block in a function and reuse it in `clear_chunk_data_and_headers`

**Review by Trisfald — APPROVED** [review]

File: chain/chain/src/tests/chunk_producers.rs [hunk 2]
```diff
@@ -431,6 +433,210 @@ mod tests {
...
+        store_update.commit().unwrap();
+
+        for anchor in [blocks[3].hash(), blocks[4].hash(), blocks[5].hash()] {
+            for shard_id in shard_layout.shard_ids() {
+                assert!(
+                    row(&chain, anchor, shard_id).is_none(),
+                    "undo-block must delete ChunkProducers for {anchor}"
+                );
+            }
+        }
```
Comment (staffik) [inline chain/chain/src/tests/chunk_producers.rs:654]: Could you check if appending this to `test_chunk_producers_garbage_collected_on_undo_block` fails:

  ```rust
  chain
      .sync_block_headers(blocks[3..=5].iter().map(|b| b.header().clone().into()).collect())
      .unwrap();
  for shard_id in shard_layout.shard_ids() {
      assert!(
          row(&chain, blocks[4].hash(), shard_id).is_some(),
          "header-only anchor's row must be re-seeded on re-sync",
      );
  }
  ```
looks like re-seeding is skipped when `BlockInfo` already exists, and we only delete `BlockInfo` for the body head, so the header-only ones above it stay missing after re-sync.

**Review by staffik — COMMENTED** [review]
.

File: chain/chain/src/tests/chunk_producers.rs (hunk 2 above)
Comment (stedfn) [reply chain/chain/src/tests/chunk_producers.rs:654]: nice catch, I did a change in [096d641](https://github.com/near/nearcore/pull/16023/commits/096d6413c93cd7190ac526006763c0f396a15b39)

File: chain/chain/src/garbage_collection.rs [hunk 3]
```diff
@@ -502,6 +502,20 @@ impl ChainStore {
...
+        let cp_keys: Vec<Box<[u8]>> = self
+            .store()
+            .iter_prefix(DBCol::ChunkProducers, header_hash.as_bytes())
+            .map(|(key, _)| key)
+            .collect();
+        for cp_key in cp_keys {
+            self.gc_col(DBCol::ChunkProducers, &cp_key);
+        }
+    }
+
```
Comment (staffik) [inline chain/chain/src/garbage_collection.rs:520]: nit: is this undo-block only? it reads like a general gc helper, took me a bit to work out why CPs get deleted in a few places. maybe a rename or a one-line doc? wdyt

File: chain/chain/src/garbage_collection.rs [hunk 4]
```diff
@@ -902,6 +933,11 @@ impl<'a> ChainStoreUpdate<'a> {
         self.gc_col(DBCol::BlockRefCount, block_hash.as_bytes());
         self.gc_outcomes(&block);
         self.gc_col(DBCol::BlockInfo, block_hash.as_bytes());
+        // ChunkProducers mirrors BlockInfo: both are seeded together in
+        // record_block_info_impl and must be deleted together. The column is
```
Comment (staffik) [inline chain/chain/src/garbage_collection.rs]: nit: "must be deleted together" reads a bit strong, `clear_chunk_data_and_headers` deletes CP rows without BlockInfo. wdyt

File: chain/chain/src/tests/chunk_producers.rs [hunk 5]
```diff
@@ -431,6 +433,240 @@ mod tests {
...
+            for shard_id in shard_layout.shard_ids() {
+                assert!(
+                    row(&chain, anchor, shard_id).is_some(),
+                    "undo-block must retain ChunkProducers for header-only anchor {anchor}"
+                );
+            }
+        }
+
+        // Reviewer's re-sync: re-syncing the headers must keep the header-only rows valid. This
+        // is cache-independent (the rows were never deleted). We deliberately do not assert the
```
Comment (staffik) [inline chain/chain/src/tests/chunk_producers.rs]: nit: a fresh EpochManager over the same store starts with an empty cache, so this could be a real assert instead of a comment. worth doing? wdyt

File: chain/chain/src/garbage_collection.rs [hunk 6]
```diff
@@ -502,6 +502,20 @@ impl ChainStore {
 }
 
 impl<'a> ChainStoreUpdate<'a> {
+    /// GC every ChunkProducers row anchored at `header_hash`. Rows are keyed by
+    /// (block_hash, shard_id); the header hash prefixes all shards' rows.
+    #[cfg(feature = "nightly")]
+    fn gc_chunk_producers_for_header(&mut self, header_hash: &CryptoHash) {
```
Comment (staffik) [inline chain/chain/src/garbage_collection.rs]: nit: setup() is 1-shard, so the scan always collects a single key here. would a multi-shard case be worth adding? wdyt

File: chain/chain/src/garbage_collection.rs [hunk 7]
```diff
@@ -686,6 +711,12 @@ impl<'a> ChainStoreUpdate<'a> {
         for shard_uid in self.get_shard_uids_to_gc(epoch_manager, &block_hash) {
             let block_shard_uid = get_block_shard_uid(&block_hash, &shard_uid);
             self.gc_col(DBCol::ChunkExtra, &block_shard_uid);
+            // ChunkProducers is keyed by ShardId (not ShardUId); insert-only, nightly-only.
+            #[cfg(feature = "nightly")]
+            self.gc_col(
+                DBCol::ChunkProducers,
```
Comment (staffik) [inline chain/chain/src/garbage_collection.rs]: nit: nothing seems to drive clear_forks_data over a fork block with a body. fine today since this delete is shared with Canonical, but maybe worth an assert? wdyt

File: test-loop-tests/src/tests/chunk_producers.rs [hunk 8]
```diff
@@ -304,3 +304,198 @@ mod cold_storage_tests {
...
+    use near_primitives::version::{PROTOCOL_VERSION, ProtocolFeature};
+    use near_store::DBCol;
+    use std::collections::BTreeMap;
+    use std::sync::Arc;
+
+    const EPOCH_LENGTH: u64 = 5;
+    const GC_NUM_EPOCHS_TO_KEEP: u64 = 3;
+
+    #[test]
+    #[cfg_attr(feature = "protocol_feature_spice", ignore)]
```
Comment (staffik) [inline test-loop-tests/src/tests/chunk_producers.rs:334]: nit: worth the TODO(spice-test) marker here too, like the other two spice-ignored tests in this file?

**Review by staffik — APPROVED** [review]
remaining comments, feel free to ignore

File: chain/chain/src/garbage_collection.rs (hunk 3 above)
Comment (stedfn) [reply chain/chain/src/garbage_collection.rs:520]: renamed to `gc_chunk_producers_for_block`

File: chain/chain/src/tests/chunk_producers.rs (hunk 5 above)
Comment (stedfn) [reply chain/chain/src/tests/chunk_producers.rs]: left as is. A fresh manager needs `genesis.config`, which `setup()` does not expose, so it means duplicating the config or touching the shared helper. Empty manager seeding is already covered by other tests

File: chain/chain/src/garbage_collection.rs (hunk 6 above)
Comment (stedfn) [reply chain/chain/src/garbage_collection.rs]: I will skip, the `test_chunk_producers_garbage_collected_at_reshard_boundary` test already has the multi-shard case end to end

File: chain/chain/src/garbage_collection.rs (hunk 7 above)
Comment (stedfn) [reply chain/chain/src/garbage_collection.rs]: added an extra test
